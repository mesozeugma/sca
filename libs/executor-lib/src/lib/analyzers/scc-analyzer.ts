import { z } from 'zod';

import { AnalysisOutput } from '../analysis/analysis-output';
import { OpensearchIndex } from '../opensearch/opensearch-index';
import { Utils } from '../utils';

import { Analyzer } from './analyzer';

const LineCountStat = z
  .object({
    Name: z.string().toLowerCase(),
    Blank: z.number().int(),
    Code: z.number().int(),
    Comment: z.number().int(),
    Count: z.number().int(),
    Lines: z.number().int(),
  })
  .strip()
  .transform((stat) => ({
    language: stat.Name,
    blankLineCount: stat.Blank,
    commentRatio: stat.Comment / (stat.Comment + stat.Code),
    codeLineCount: stat.Code,
    commentLineCount: stat.Comment,
    fileCount: stat.Count,
    lineCount: stat.Lines,
  }));
type LineCountStat = z.infer<typeof LineCountStat>;
const sccOutput = z.array(LineCountStat);

export class SccAnalyzer implements Analyzer {
  private async scc(repositoryPath: string): Promise<LineCountStat> {
    if (!repositoryPath) {
      throw Error('Repository path not set');
    }

    const { stdout } = await Utils.executeCommand('scc', [
      '--format=json',
      '--sort=code',
      repositoryPath,
    ]);
    const parseResult = await sccOutput.parseAsync(JSON.parse(stdout));
    const primaryLanguageStat = parseResult.find(({ language }) =>
      ['java', 'python'].includes(language)
    );
    if (primaryLanguageStat !== undefined) {
      return primaryLanguageStat;
    }
    throw new Error('No countable languages in repository');
  }

  async readRepo(repositoryPath: string) {
    const result = await this.scc(repositoryPath);
    return [
      new AnalysisOutput(
        OpensearchIndex.ANALYSIS_RESULT_PRIMARY_LINE_COUNTS,
        result
      ),
    ];
  }
}
