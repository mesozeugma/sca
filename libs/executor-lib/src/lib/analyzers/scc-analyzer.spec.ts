import { SccAnalyzer } from './scc-analyzer';

test('SccAnalyzer', async () => {
  const analyzer = new SccAnalyzer();
  const result = await analyzer.readRepo(
    './libs/executor-lib/code-samples/java/'
  );
  expect(result).toMatchSnapshot();
});
