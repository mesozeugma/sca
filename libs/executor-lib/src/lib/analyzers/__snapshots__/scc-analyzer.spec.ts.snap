// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SccAnalyzer 1`] = `
[
  AnalysisOutput {
    "data": {
      "blankLineCount": 5,
      "codeLineCount": 18,
      "commentLineCount": 1,
      "commentRatio": 0.05263157894736842,
      "fileCount": 2,
      "language": "java",
      "lineCount": 24,
    },
    "index": "public_sca_analysis_result_primary_line_counts",
  },
]
`;
